### ISSUE TITLE
obsoletion corner cases handling idea

### ISSUE THREAD

**Issue by nirik** [fedora-infra/bodhi#1266]
Right now there's a few corner cases for obsoleting updates:

* If the older update is locked/in a push
* If the older update has multiple packages attached to it
* If the older update is submitted to stable?

I'd like to suggest we handle these by a simple rule: If any of the above are the case, we simply reject the new update being created and tell the submittor that there is already an older update XYZ and it cannot be obsoleted, and please edit that update.

Thoughts?

**Comment by tyll** [comment]
IMHO it would be better to not block or waste the contributors work.

1. if the older update is locked/in a push (I guess to testing), then the new update could be put in a waiting state and once the push is completed, Bodhi could obsolete the old update
2. if the older update has multiple packages attached to it, bodhi should offer to update the older update and replace the builds and make it easy to add the other information that might be missing (new bugs, update notes)
3. if the older update is submitted to stable, Bodhi could allow the newer update to be submitted to testing, I do not see the issue here

**Comment by bowlofeggs** [comment]
@puiterwijk has been advocating to just have Bodhi be mean and refuse to create new updates when packages are found in other active updates. It would be less fancy, but it will also solve a few difficult-to-solve issues.

Even the request: stable state can be a problem, because it is possible for a packager to click "unpush" on it, which could get it back into a state of conflict.

**Comment by nirik** [comment]
Well, I think just denying it is the easy solution we might be able to deploy sooner rather than later. 

Some of those other solutions could be things we implement down the road, but the problem is that it's hard to know what the submittor intends, and there's tons of weird corner cases because sometimes maintainer A submits something and maintainer B is doing something else. Rejecting the second one at least gives maintainer B incentive to talk to maintainer A, find out whats going on and hopefully do the right thing. :grin: 

Some more corner cases: If you queue the new update when an older one is locked, what happens if you submit 10? Which one goes in? all of them? The last one? If the new update has 100 things in it (say a new kf5 version, or a rebuild of a bunch of packages with a fixed compiler) and there's 5 seperate updates already in existance, which should it let you replace builds on? all of them?

**Comment by tyll** [comment]
My opinion is that the machines should server the humans and not the other way around. If the maintainers intention is unknown, then they need to be kept in control. If automatically doing something with updates in the waiting state, they could just stay there and wait for the maintainer to submit them to testing. And if they are too old they could be garbage collected.

And if there are multiple new updates, you can ask the maintainer to merge/replace or queue the new update. But I guess usually merge or replace make sense here. And about the new update with 100 things in it, do you believe it will make the maintainer happier when they click this together on the web interface and then it will be rejected and there is no way to store this? They will probably use the CLI for this amount, but maybe not for 5-10 builds. And it is clear that denying the update is not want any maintainer wants, because they want the updates to get out somehow and Bodhi should support them to do the right thing.

**Comment by mohanboddu** [comment]
Another ticket filed with fedora releng https://pagure.io/releng/issue/10662

**Comment by AdamWill** [comment]
Coming back to this in the wake of https://pagure.io/releng/issue/13058 : I have to disagree with @tyll and agree with @nirik . Different builds of the same package being in multiple in-flight updates for the same release (excepting 'pending stable push', I *guess*...but even that can be troublesome with freezes) is not a sane and supportable state. There are so many ways it can go wrong.

We should disallow it. But on the web UI side we should do it as gracefully as possible, in a way that *doesn't* lose all the work the submitter has put into building the package list and the description - we should keep all that state and allow them to just take the problematic build out of the update and then hit the submit button again, or whatever they want to do.

**Comment by tyll** [comment]
How would preventing the second update help with the gstreamer issue @AdamWill? If the old (broken) update blocks submission of the new, fixed update, there will only be a bigger delay until the fixed update reaches the users.

**Comment by AdamWill** [comment]
@tyll there wasn't an "old broken update" and a "new fixed update". There was an update containing a build of gstreamer1-plugins-good against the then-stable Qt version, and an update containing a new Qt version with gstreamer1-plugins-good rebuilt against it. The pre-qt-bump update was perfectly cromulent *on its own terms*.

We should not have allowed the Qt bump update to be created with a gstreamer1-plugins-good build in it while there was an existing update with gstreamer1-plugins-good built against the old Qt. That was the problem. The qt update should either have waited until the existing update cleared through, or merged with it (i.e. the existing gstreamer1-plugins-good update could have become the qt bump update too).

**Comment by Conan-Kudo** [comment]
That also gets into merging side tags, which is currently not a thing either (though it really should be).

**Comment by AdamWill** [comment]
yeah, although I think you can at least manually untag the builds from one side tag and tag them into the other, if both builds use side tags? I don't think I've tried it, but seems like it ought to be possible...maybe only koji admins can do it or something...

**Comment by nirik** [comment]
You absoluetely can. ;) 

Rejecting at the bodhi level punts this problem to maintainers to sort out. There's a number of ways they could do so... merge things, do more builds, wait, etc. But rejecting it here means they do have to do something, they can't just ignore this and break everything.
